modin/engines/ray/generic/io.py

```diff
@@ -44,3 +46,86 @@ def func(df):
         result = qc._modin_frame._apply_full_axis(1, func, new_index=[], new_columns=[])
         # blocking operation
         result.to_pandas()
+
+    @staticmethod
+    def _to_csv_check_support(kwargs):
+        path_or_buf = kwargs["path_or_buf"]
+        compression = kwargs["compression"]
+        if not isinstance(path_or_buf, str):
+            return False
+
+        # case when the pointer is placed at the beginning of the file.
+        if "r" in kwargs["mode"] and "+" in kwargs["mode"]:
+            return False
+
+        # encodings with BOM don't support;
+        # instead of one mark in result bytes we will have them by the number of partitions
+        # so we should fallback in pandas for `utf-16`, `utf-32` with all aliases, in instance
+        # (`utf_32_be`, `utf_16_le` and so on)
+        if kwargs["encoding"] is not None:
+            encoding = kwargs["encoding"].lower()
+            if "u" in encoding or "utf" in encoding:
+                if "16" in encoding or "32" in encoding:
+                    return False
+
+        if compression is None or not compression == "infer":
+            return False
+
+        if any((path_or_buf.endswith(ext) for ext in [".gz", ".bz2", ".zip", ".xz"])):
+            return False
+
+        return True
+
+    @classmethod
+    def to_csv(cls, qc, **kwargs):
+        if not cls._to_csv_check_support(kwargs):
+            return BaseIO.to_csv(qc, **kwargs)
+
+        # The partition id will be added to the queue, for which the moment
+        # of writing to the file has come
+        queue = Queue(maxsize=1)
+
+        def func(df, **kw):
+            if kw["partition_idx"] != 0:
+                if "w" in kwargs["mode"]:
+                    kwargs["mode"] = kwargs["mode"].replace("w", "a")
+                kwargs["header"] = False
+
+            path_or_buf = kwargs["path_or_buf"]
+            is_binary = "b" in kwargs["mode"]
+            if is_binary:
+                kwargs["path_or_buf"] = io.BytesIO()
+            else:
+                kwargs["path_or_buf"] = io.StringIO()
+
+            df.to_csv(**kwargs)
+            content = kwargs["path_or_buf"].getvalue()
+
+            while True:
+                get_value = queue.get(block=True)
+                if get_value == kw["partition_idx"]:
+                    break
+                queue.put(get_value)
+
+            open_kwargs = {"mode": kwargs["mode"]}
+            if not is_binary:
+                open_kwargs["newline"] = ""
+
+            with open(path_or_buf, **open_kwargs) as _f:
+                _f.write(content)
+
+            queue.put(get_value + 1)
+
+            return pandas.DataFrame()
+
+        queue.put(0)
+        result = qc._modin_frame.broadcast_apply_full_axis(
```

Comment:
Do we really need to use `broadcast_apply`? I think the regular `map_axis_partitions` would be better here.